earthmover/nodes/destination.py

```diff
@@ -103,20 +103,23 @@ def execute(self, **kwargs):
         # Verify the output directory exists.
         os.makedirs(os.path.dirname(self.file), exist_ok=True)
 
-        # Write the optional header, the JSON lines as CSV (for performance), and the optional footer.
-        self.data.to_csv(
-            filename=self.file, single_file=True, mode='wt', index=False,
-            header=[self.header] if self.header else False,  # We must write the header directly due to aforementioned bug.
-            escapechar="\x01", sep="\x02", quoting=csv.QUOTE_NONE,  # Pretend to be CSV to improve performance
-        )
+        # Write the optional header, each line, and the optional footer.
+        with open(self.file, 'w+', encoding='utf-8') as fp:
+            if self.header:
+                fp.write(self.header)
+
+            self.data.apply(self.write_row, meta=pd.Series('str'), fp=fp).compute()
 
-        if self.footer:
-            with open(self.file, 'a', encoding='utf-8') as fp:
+            if self.footer:
                 fp.write(self.footer)
 
         self.logger.debug(f"output `{self.file}` written")
         self.size = os.path.getsize(self.file)
 
+    def write_row(self, row: pd.Series, fp):
+        fp.write(row + "\n")
+        return "" # this wipes out data in the dataframe after it's written, which should save some memory
```

Comment:
FYI Tom, empty strings in Python are 49 bytes apiece, but `None`s are 16 bytes. I don't know how the eventual PyArrow strings will affect this. Technically, by not having a return, there is an implicit `return None`.

Reply:
I updated this, and I left the explicit `return None` with the comment explaining why.